provisioning/miniprov/hare_mp/cdf.py

```diff
@@ -132,37 +132,47 @@ def _validate_pool(self, pool: PoolHandle) -> None:
                 f'device count ({data_devices_count}) must be not '
                 f'less than N+K+S ({min_width})')
 
-    def _add_pool(self, pool: PoolHandle, out_list: List[PoolDesc]) -> None:
+    # Current formula is as follows
+    # Disk or device failure = K(parity)
+    # Controller or CVG failure = min(K, cvg per node * number of nodes) 
+    # Node failures = floor(K/ceil((N+K+S)/ number of nodes))
+    def _calculate_allowed_failure(self, layout: Layout) -> AllowedFailures:
         conf = self.provider
-        layout = self._get_layout(pool)
-        if not layout:
-            return
-        (cid, pool_type, i) = pool.tuple()
-        storage_set_name = conf.get(f'cluster>{cid}>storage_set[{i}]>name')
-        pool_name = f'{storage_set_name}__{pool_type}'
-
-        ctrl_failure_allowed = layout.parity
 
         machine_id = conf.get_machine_id()
         node_count = len(conf.get_storage_set_nodes())
         cvg_per_node = int(conf.get(
             f'server_node>{machine_id}>storage>cvg_count'))
 
-        cvg_per_storage_set = cvg_per_node * node_count
-
-        if layout.data + layout.parity + layout.spare == 0:
+        total_unit = layout.data + layout.parity + layout.spare
+        if total_unit == 0:
             raise RuntimeError('All layout parameters are 0')
 
-        ratio = floor(cvg_per_storage_set / (
-                      layout.data + layout.parity + layout.spare))
+        enc_failure_allowed_FP = layout.parity / ceil(total_unit / node_count)
+        enc_failure_allowed = floor(enc_failure_allowed_FP)
 
-        if ratio == 0:
-            ctrl_failure_allowed = 0
+        cvg_per_storage_set = cvg_per_node * node_count
+        if cvg_per_storage_set < layout.parity:
+            ctrl_failure_allowed = cvg_per_storage_set
         else:
             ctrl_failure_allowed = layout.parity
```

Comment:
This will not work for 8 + 2 + 0, 3 nodes, 1 cvg per node and 4 devices per cvg.
encl_failures = 0 (correct)
ctrl_failures = 2 (wrong) we cannot support a ctrl failure as 4 devices will fail and per controller there will be atleast 3 parity group units. So we can ideally support just 2 drive failures. 
We need a validation function for this. There is one in the Motr, I think we need to port it to Hare for validations.

Reply:
I had quick call with Madhavrao for this.
@madhavemuri Please let us know once you are done with modification of formula on confluence page.

Reply:
https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/127893680/CORTX+Durability+Configuration
I have updated the page, which should take care of above case as well.

Reply:
@madhavemuri, Okay, the change makes sense.
@SwapnilGaonkar7 can you please make the change in the formula?

Reply:
@mssawant I have modified formula accordingly

Reply:
@SwapnilGaonkar7 as per the formula,
```
min (A*#NodeFailure, K)
```
you need take the minimum of cvg failures or K, I don't see that here.

Reply:
temp = cvg_per_node * enc_failure_allowed
        if temp < layout.parity:
            ctrl_failure_allowed = temp
        else:
            ctrl_failure_allowed = layout.parity

This code should handle above condition where A is cvg_per_node